### PR TITLE
MXMNetSphericalBasis Layer

### PR DESCRIPTION
**Pull request by riya-singh28**
## Description

This layer combines radial basis functions (RBF) and spherical harmonic functions to transform input distances and angles.

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by riya-singh28** [comment]
image

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -4934,3 +4936,97 @@ def forward(self, one_electron: torch.Tensor, two_electron: torch.Tensor):
...
+            The number of radial basis functions to use. These functions capture information about pairwise distances between atoms.
+        cutoff: float, optional (default 5.0)
+            The cutoff distance for the radial basis functions. It specifies the distance beyond which the interactions are ignored.
+        envelope_exponent: int, optional (default 5)
+            The exponent for the envelope function. It controls the degree of damping for the radial basis functions.
+        """
+        super(MXMNetSphericalBasisLayer, self).__init__()
+
+        assert num_radial <= 64
+        self.num_spherical = num_spherical
```
Comment (ARY2260) [inline deepchem/models/torch_models/layers.py:4977]: missing type hinting

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -4934,3 +4936,97 @@ def forward(self, one_electron: torch.Tensor, two_electron: torch.Tensor):
...
+        self.num_radial = num_radial
+        self.cutoff = cutoff
+        self.envelope: _MXMNetEnvelope = _MXMNetEnvelope(envelope_exponent)
+
+        bessel_forms: List = bessel_basis(num_spherical, num_radial)
+        sph_harm_forms: List[List[str]] = real_sph_harm(num_spherical)
+        self.sph_funcs: List = []
+        self.bessel_funcs: List = []
+
+        x, theta = sym.symbols('x theta')
```
Comment (ARY2260) [inline deepchem/models/torch_models/layers.py:4987]: type hinting missing here

File: deepchem/models/torch_models/layers.py [hunk 3]
```diff
@@ -4934,3 +4936,97 @@ def forward(self, one_electron: torch.Tensor, two_electron: torch.Tensor):
...
+        rbf: torch.Tensor = torch.stack([f(dist) for f in self.bessel_funcs],
+                                        dim=1)
+        rbf = self.envelope(dist).unsqueeze(-1) * rbf
+
+        cbf: torch.Tensor = torch.stack([f(angle) for f in self.sph_funcs],
+                                        dim=1)
+        n: int = self.num_spherical
+        k: int = self.num_radial
+
+        output = (rbf[idx_kj].view(-1, n, k) * cbf.view(-1, n, 1)).view(
```
Comment (ARY2260) [inline deepchem/models/torch_models/layers.py:5030]: missing type hint

File: deepchem/models/torch_models/layers.py [hunk 4]
```diff
@@ -4934,3 +4936,97 @@ def forward(self, one_electron: torch.Tensor, two_electron: torch.Tensor):
             two_electron = two_electron_tmp
 
         return one_electron, two_electron
+
+
+class MXMNetSphericalBasisLayer(torch.nn.Module):
+    """This layer combines radial basis functions (RBF) and spherical harmonic functions
```
Comment (ARY2260) [inline deepchem/models/torch_models/layers.py:4942]: Please describe the layer in detail and mention the references

**Review by ARY2260 — CHANGES_REQUESTED** [review]
I have added few comments and suggestions

File: deepchem/models/torch_models/layers.py (hunk 1 above)
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4977]: I'll add it here.

File: deepchem/models/torch_models/layers.py (hunk 4 above)
Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4942]: Okay I'll do it.

**Review by ARY2260 — APPROVED** [review]

**Comment by ARY2260** [comment]
LGTM

**Comment by ARY2260** [comment]
LGTM, @rbharath can you please take a look at this PR.

File: deepchem/models/torch_models/layers.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import math
 from math import pi as PI
 import numpy as np
+import sympy as sym
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:4]: We need to add this to the appropriate sections of the docs. Try reading through and seeing where we list requirements

Comment (ARY2260) [reply deepchem/models/torch_models/layers.py:4]: https://github.com/riya-singh28/deepchem/blob/9f591b0ebcc0a270c480ae397cb821e950c34efc/requirements/env_common.yml

might need to add sympy here as a hard depencency

**Comment by riya-singh28** [comment]
@rbharath Can you please do a final review pass on this PR?

File: deepchem/models/torch_models/layers.py (hunk 5 above)
Comment (rbharath) [reply deepchem/models/torch_models/layers.py:4]: Still need to add it here

Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4]: Had a discussion on this during OH hours and made the suggested changes.

Comment (riya-singh28) [reply deepchem/models/torch_models/layers.py:4]: The sympy lib is installed before scipy directly in setup.py to ensure the fixed version of scipy after the sympy is installed as sympy installs an older version of scipy.

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**